File: src/main/java/com/viartemev/requestmapper/RequestMappingItem.java
package com.viartemev.requestmapper;

import com.intellij.navigation.ItemPresentation;
import com.intellij.navigation.NavigationItem;
import com.intellij.pom.Navigatable;
import com.intellij.psi.PsiClass;
import com.intellij.psi.PsiElement;
import com.intellij.psi.PsiMethod;

import javax.swing.*;
import java.util.Optional;

public class RequestMappingItem implements NavigationItem {
    private final PsiElement psiElement;
    private final String urlPath;
    private final Alias alias;
    private final RequestMappingItemPresentation mItemPresentation;

    public RequestMappingItem(PsiElement psiElement, String urlPath, String requestMethod) {
        this.psiElement = psiElement;
        this.urlPath = urlPath;
        this.alias = new Alias(psiElement, requestMethod);
        this.mItemPresentation = new RequestMappingItemPresentation();
    }

    public PsiElement getPsiElement() {
        return this.psiElement;
    }

    @Override
    public String getName() {
        return this.urlPath;
    }

    @Override
    public ItemPresentation getPresentation() {
        return this.mItemPresentation;
    }


    @Override
    public void navigate(boolean requestFocus) {
        navigationElement().ifPresent(navigatable -> navigatable.navigate(requestFocus));
    }

    @Override
    public boolean canNavigate() {
        return navigationElement().map(Navigatable::canNavigate).orElse(false);
    }

    private Optional<Navigatable> navigationElement() {
        PsiElement navigationElement;
        if (getPsiElement() != null
                && (navigationElement = getPsiElement().getNavigationElement()) != null
                && navigationElement instanceof Navigatable) {
            return Optional.of((Navigatable) navigationElement);
        } else {
            return Optional.empty();
        }
    }

    @Override
    public boolean canNavigateToSource() {
        return true;
    }

    private class RequestMappingItemPresentation implements ItemPresentation {

        private RequestMappingItemPresentation() {
            //empty ctor
        }

        @Override
        public String getPresentableText() {
            return RequestMappingItem.this.urlPath;
        }

        @Override
        public String getLocationString() {
            return RequestMappingItem.this.alias.getAlias();
        }

        @Override
        public Icon getIcon(boolean b) {
            return RequestMapperIcons.SEARCH;
        }
    }

    private class Alias {
        private final PsiElement psiElement;
        private final String requestMethod;

        public Alias(PsiElement psiElement, String requestMethod) {
            this.psiElement = psiElement;
            this.requestMethod = requestMethod;
        }

        public String getAlias() {
            if (psiElement instanceof PsiMethod) {
                PsiMethod psiElement = (PsiMethod) this.psiElement;
                PsiClass clss;
                PsiMethod method;
                return requestMethod + " " + ((clss = (method = psiElement).getContainingClass()) == null ? "" : clss.getName() + ".") + method.getName();
            } else if (psiElement instanceof PsiClass) {
                return ((PsiClass) psiElement).getName();
            } else {
                return "undefined";
            }
        }
    }

}

File: src/main/kotlin/com/viartemev/requestmapper/RequestMappingItem.kt
package com.viartemev.requestmapper

import com.intellij.navigation.ItemPresentation
import com.intellij.navigation.NavigationItem
import com.intellij.pom.Navigatable
import com.intellij.psi.PsiClass
import com.intellij.psi.PsiElement
import com.intellij.psi.PsiMethod

class RequestMappingItem(val psiElement: PsiElement, private val urlPath: String, private val requestMethod: String) : NavigationItem {

    private val navigationElement = psiElement.navigationElement as? Navigatable

    override fun getName(): String = this.requestMethod + " " + this.urlPath

    override fun getPresentation(): ItemPresentation = RequestMappingItemPresentation()

    override fun navigate(requestFocus: Boolean) = navigationElement?.navigate(requestFocus) ?: Unit

    override fun canNavigate(): Boolean = navigationElement?.canNavigate() ?: false

    override fun canNavigateToSource(): Boolean = true

    override fun toString(): String {
        return "RequestMappingItem(urlPath='$urlPath', requestMethod='$requestMethod')"
    }

    private inner class RequestMappingItemPresentation : ItemPresentation {

        override fun getPresentableText() = this@RequestMappingItem.requestMethod + " " + this@RequestMappingItem.urlPath

        override fun getLocationString(): String {
            val psiElement = this@RequestMappingItem.psiElement
            val fileName = psiElement.containingFile?.name
            return when (psiElement) {
                is PsiMethod -> (psiElement.containingClass?.name ?: fileName ?: "unknownFile") + "." + psiElement.name
                is PsiClass -> psiElement.name ?: fileName ?: "unknownFile"
                else -> "unknownLocation"
            }
        }

        override fun getIcon(b: Boolean) = RequestMapperIcons.SEARCH
    }
}
